server/src/main/java/io/druid/client/DirectDruidClient.java

```diff
@@ -444,6 +455,33 @@ private void checkTotalBytesLimit(long bytes)
             throw new RE(msg);
           }
         }
+
+        private long getBufferCapacity()
+        {
+          return maxBufferSizeBytes - totalBufferedBytes.get();
+        }
+
+        private void checkBufferCapacity(long bytes, long timeoutMs)
```

Comment:
It looks like this method will block until buffer space becomes available. We should make sure it's not running in any threads where that would be problematic.